### PR TITLE
support for resuming magnet links

### PR DESCRIPTION
**Pull request by sweetppro**
magnet links are now saved as a text file the `Torrents` folder as `.magnet` files.

macOS support is added in [PR #2686](https://github.com/transmission/transmission/pull/2686):

### CODE REVIEW

**Comment by sweetppro** [comment]
no idea why my PRs keep failing checks...
:(

File: libtransmission/torrent.cc [hunk 1]
```diff
@@ -767,17 +767,35 @@ static void torrentInit(tr_torrent* tor, tr_ctor const* ctor)
 
     tr_sessionAddTorrent(session, tor);
 
-    // if we don't have a local .torrent file already, assume the torrent is new
-    auto const filename = tor->torrentFile();
+    // if we don't have a local .torrent or .magent file already, assume the torrent is new
```
Comment (ckerr) [inline libtransmission/torrent.cc:770]:
```suggestion
    // if we don't have a local .torrent or .magnet file already, assume the torrent is new
```

File: libtransmission/session.cc [hunk 2]
```diff
@@ -2041,13 +2041,25 @@ static void sessionLoadTorrents(void* vdata)
...
+                    
+            // is a magnet link?
+            if (!tr_ctorSetMetainfoFromFile(data->ctor, path, nullptr))
+            {
+                auto buf = std::vector<char>{};
+                if (!tr_loadFile(buf, path))
+                {
+                    continue;
+                }
+                auto const magnet_link = std::string{ std::data(buf), std::size(buf) };
```
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/session.cc:2059]: this generates the error:
`No member named 'c_str' in 'std::string_view'`

then if I change the following line to:
`tr_ctorSetMetainfoFromMagnetLink(data->ctor, magnet_link, nullptr);`
I get the following error:
`No matching function for call to 'tr_ctorSetMetainfoFromMagnetLink'`

File: libtransmission/torrent-ctor.cc [hunk 3]
```diff
@@ -127,6 +127,21 @@ bool tr_ctorSaveContents(tr_ctor const* ctor, std::string const& filename, tr_er
...
+    TR_ASSERT(!std::empty(filename));
+    
+    if (std::empty(tor->metainfo_.magnetLink()))
+    {
+        tr_error_set(error, EINVAL, "torrent has no magnetlink to save"sv);
+        return false;
+    }
+
+    std::string magnet_link = tor->metainfo_.magnetLink();
+    return tr_saveFile(filename, std::string_view{ magnet_link }, error);
```
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/torrent-ctor.cc:142]: this generates the error:
`No member named 'magnetLink' in 'tr_torrent'`

File: libtransmission/magnet-metainfo.h [hunk 4]
```diff
@@ -61,11 +64,17 @@ class tr_magnet_metainfo
...
+        magnet_link_ = magnet_link;
+    }
 
 protected:
     tr_announce_list announce_list_;
     std::vector<std::string> webseed_urls_;
     tr_sha1_digest_t info_hash_;
     std::string info_hash_str_;
     std::string name_;
+    std::string magnet_link_;
```
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/magnet-metainfo.h:79]: yeah, we can get rid of this

File: libtransmission/session.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/session.cc:2059]: ill change to this:
```
auto const magnet_link = std::string_view{ std::data(buf), std::size(buf) };
tr_ctorSetMetainfoFromMagnetLink(data->ctor, std::string{ magnet_link }, nullptr);
```

File: libtransmission/torrent-ctor.cc [hunk 5]
```diff
@@ -127,6 +127,21 @@ bool tr_ctorSaveContents(tr_ctor const* ctor, std::string const& filename, tr_er
...
 
+bool tr_ctorSaveMagnetContents(tr_torrent* tor, std::string const& filename, tr_error** error)
+{
+    TR_ASSERT(tor != nullptr);
+    TR_ASSERT(!std::empty(filename));
+    
+    if (std::empty(tor->metainfo_.magnetLink()))
+    {
+        tr_error_set(error, EINVAL, "torrent has no magnetlink to save"sv);
+        return false;
```
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/torrent-ctor.cc:138]: ive removed it

File: libtransmission/torrent-ctor.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (sweetppro) [reply libtransmission/torrent-ctor.cc]: I’ll look into this when I get hone

File: libtransmission/session.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ckerr) [reply libtransmission/session.cc]: Oh of course you're right; tr_ctorSetMetainfoFromMagnetLink requires a C pointer. My original comment was wrong.

**Comment by ckerr** [comment]
> no idea why my PRs keep failing checks... :(

There is a known issue in CI in the rename / move sanity tests, that is my bug -- it's just taking longer to fix than I expected. If that's the error you're getting, feel free to ignore it

File: libtransmission/magnet-metainfo.cc [hunk 6]
```diff
@@ -180,8 +180,6 @@ std::string tr_magnet_metainfo::magnet() const
 
 bool tr_magnet_metainfo::parseMagnet(std::string_view magnet_link, tr_error** error)
 {
-    magnet_link = tr_strvStrip(magnet_link);
-
```
Comment (ckerr) [inline libtransmission/magnet-metainfo.cc:184]: This shouldn't've been removed -- this line lets magnet links that accidentally have leading or trailing whitespace still get parsed correctly

File: libtransmission/magnet-metainfo.h [hunk 7]
```diff
@@ -22,11 +22,12 @@ class tr_magnet_metainfo
     bool parseMagnet(std::string_view magnet_link, tr_error** error = nullptr);
 
     std::string magnet() const;
-
+
```
Comment (ckerr) [inline libtransmission/magnet-metainfo.h:25]: _(nit)_ accidentally-introduced whitespace, here and a few lines down on line 30

File: libtransmission/session.cc [hunk 8]
```diff
@@ -2041,13 +2041,25 @@ static void sessionLoadTorrents(void* vdata)
...
+            // is a magnet link?
+            if (!tr_ctorSetMetainfoFromFile(data->ctor, path, nullptr))
+            {
+                auto buf = std::vector<char>{};
+                if (!tr_loadFile(buf, path))
+                {
+                    continue;
+                }
+                auto const magnet_link = std::string_view{ std::data(buf), std::size(buf) };
+                tr_ctorSetMetainfoFromMagnetLink(data->ctor, std::string{ magnet_link }, nullptr);
```
Comment (ckerr) [inline libtransmission/session.cc:2060]:
```
if (auto buf = std::vector<char>{}; tr_loadFile(buf, path))
{
    tr_ctorSetMetainfoFromMagnetLink(data->ctor, std::string{buf}.c_str(), nullptr);
}
```

File: libtransmission/torrent-ctor.cc [hunk 9]
```diff
@@ -127,6 +127,21 @@ bool tr_ctorSaveContents(tr_ctor const* ctor, std::string const& filename, tr_er
...
+    TR_ASSERT(!std::empty(filename));
+    
+    if (std::empty(tor->metainfo_.magnet()))
+    {
+        tr_error_set(error, EINVAL, "torrent has no magnetlink to save"sv);
+        return false;
+    }
+
+    std::string magnet_link = tor->metainfo_.magnet().c_str();
+    return tr_saveFile(filename, std::string_view{ magnet_link }, error);
```
Comment (ckerr) [inline libtransmission/torrent-ctor.cc:142]: We can avoid generating the magnet link twice

```suggestion
    auto const magnet = tor->magnet();
    
    if (std::empty(magnet))
    {
        tr_error_set(error, EINVAL, "torrent has no magnetlink to save"sv);
        return false;
    }

    return tr_saveFile(filename, magnet, error);
```

**Review by ckerr — COMMENTED** [review]
This is great stuff. Thank you for doing it!

File: libtransmission/magnet-metainfo.cc (hunk 6 above)
Comment (sweetppro) [reply libtransmission/magnet-metainfo.cc:184]: Oope

File: libtransmission/magnet-metainfo.h (hunk 7 above)
Comment (sweetppro) [reply libtransmission/magnet-metainfo.h:25]: Yup I noticed tgis

**Review by ckerr — COMMENTED** [review]
LGTM. The PR needs you to run `./code-style.sh` in the top-level directory to run clang-format to make the code style CI bot happy, but that's the only remaining issue that I see.

**Comment by sweetppro** [comment]
> LGTM. The PR needs you to run `./code-style.sh` in the top-level directory to run clang-format to make the code style CI bot happy, but that's the only remaining issue that I see.

hmmm,
its not working for me...
Im running macOS 12.2.1

**Comment by ckerr** [comment]
no worries, I can get it

**Review by ckerr — APPROVED** [review]

**Pull request merged**